Date: Fri, 23 Jun 2023 18:01:06 +0000
Subject: crypto: ecc - Add SP800-56A rev 3 Pair-wise Consistency check

add pair-wise consistency check in fips mode as per SP800-56 rev 3
section 5.6.2.1.4:
For an ECC key pair (d, Q): Use the private key, d, along with the generator G
and other domain parameters associated with the key pair, to compute dG
(according to the rules of elliptic-curve arithmetic). Compare the
result to the public key, Q. If dG is not equal to Q, then the
pair-wise consistency test fails.

Signed-off-by: Mahmoud Adam <mngyadam@amazon.com>

[ Since ecc_pairwise_test is only used in this function, remove it from
ecc.h and mark it as static. Also move it to be declared before it is
used. - hailmo ]

Signed-off-by: Elena Avila <ellavila@amazon.com>

diff --git a/crypto/ecc.c b/crypto/ecc.c
--- a/crypto/ecc.c
+++ b/crypto/ecc.c
@@ -1563,6 +1563,56 @@ int ecc_gen_privkey(unsigned int curve_id, unsigned int ndigits,
 }
 EXPORT_SYMBOL(ecc_gen_privkey);
 
+/**
+* SP800-56A section 5.6.2.1.4 Pair-Wise Consistency Test
+* ecc_pairwise_test() - Pair-wise Consistency test
+*
+* @curve: 		elliptic curve domain parameters
+* @private_key: 	pregenerated private key for the given curve
+* @pk: 		public key as a point
+* @ndigits: 		curve's number of digits
+*
+* Pair-wise Consistency test according to SP800-56A section 5.6.2.1.4
+*
+* Return: 0 if test is successful, -EINVAL if test is failed.
+*/
+static int ecc_pairwise_test(const struct ecc_curve *curve,
+		      const u64 *private_key,
+		      struct ecc_point *pk,
+		      unsigned int ndigits)
+{
+	u64 priv[ECC_MAX_DIGITS];
+	struct ecc_point *epk;
+	int ret;
+
+	ecc_swap_digits(private_key, priv, ndigits);
+
+	epk = ecc_alloc_point(ndigits);
+	if (!epk) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	ecc_point_mult(epk, &curve->g, priv, NULL, curve, ndigits);
+
+	/* check expected public key against the public_key */
+	if (vli_cmp(epk->x, pk->x, ndigits)) {
+		ret = -EINVAL;
+		goto err_free_point;
+	}
+
+	if (vli_cmp(epk->y, pk->y, ndigits)) {
+		ret = -EINVAL;
+		goto err_free_point;
+	}
+
+	ret = 0;
+err_free_point:
+	ecc_free_point(epk);
+err:
+	return ret;
+}
+
 int ecc_make_pub_key(unsigned int curve_id, unsigned int ndigits,
 		     const u64 *private_key, u64 *public_key)
 {
@@ -1589,6 +1639,12 @@ int ecc_make_pub_key(unsigned int curve_id, unsigned int ndigits,
 		goto err_free_point;
 	}
 
+	if (fips_enabled &&
+	    ecc_pairwise_test(curve, private_key, pk, ndigits)) {
+		fips_fail_notify();
+		panic("ecc_pairwise_test failed");
+	}
+
 	ecc_swap_digits(pk->x, public_key, ndigits);
 	ecc_swap_digits(pk->y, &public_key[ndigits], ndigits);
 
